### PR TITLE
fix[lang]: pure access analysis

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
fix https://github.com/vyperlang/vyper/issues/3894, fix https://github.com/vyperlang/vyper/issues/3830

use analysis instead of pattern matching on syntax.
update code style for the pure decorator tests

### Commit message

Commit message for the final, squashed PR. (Optional, but reviewers will appreciate it! Please see [our commit message style guide](../../master/docs/style-guide.rst#best-practices-1) for what we would ideally like to see in a commit message.)

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Review by tserg — APPROVED** [review]

**Review by fubuloubu — APPROVED** [review]

**Comment by fubuloubu** [comment]
Is this a bug or an optimization?

**Comment by cyberthirst** [comment]
this isn't right?
```python
src = """
first: public(uint256)
second: public(uint256)

@internal
@pure
def foo() -> uint256:
    return self.codesize

@deploy
def __init__():
    self.first = self.foo()


@external
def bar():
    self.second = self.foo() 
"""

c = b.loads(src)
print(c.first())
c.bar()
print(c.second())
```
outputs:
```
201
136
```

**Comment by charles-cooper** [comment]
> this isn't right?
> 
> ```python
> src = """
> first: public(uint256)
> second: public(uint256)
> 
> @internal
> @pure
> def foo() -> uint256:
>     return self.codesize
> 
> @deploy
> def __init__():
>     self.first = self.foo()
> 
> 
> @external
> def bar():
>     self.second = self.foo() 
> """
> 
> c = b.loads(src)
> print(c.first())
> c.bar()
> print(c.second())
> ```
> 
> outputs:
> 
> ```
> 201
> 136
> ```

hm yea we should probably ban codesize (it's not currently classified as a state variable)

File: tests/functional/codegen/features/decorators/test_pure.py [hunk 1]
```diff
@@ -49,58 +47,120 @@ def foo() -> int128: pure
...
+    with pytest.raises(StateAccessViolation):
+        compile_code(code)
+
+
+def test_invalid_codesize_access(get_contract):
+    code = """
+@pure
+@external
+def foo() -> uint256:
+    return msg.sender.codesize
```
Comment (cyberthirst) [inline tests/functional/codegen/features/decorators/test_pure.py:71]: `msg.sender` is itself blocked, so we don't know whether it fails to compile for the right reason?

Comment (charles-cooper) [reply tests/functional/codegen/features/decorators/test_pure.py]: nice find. fixed in https://github.com/vyperlang/vyper/pull/3895/commits/45192f67f3a596129a09d1d086e07d99cec54c17

**Comment by pcaversaccio** [comment]
This commit introduced a regression. Tracking here: https://github.com/vyperlang/vyper/issues/3902.

**Pull request merged**